endpoint_route_handler/models/ir_http.py

```diff
@@ -0,0 +1,82 @@
+# Copyright 2021 Camptocamp SA
+# @author: Simone Orsi <simone.orsi@camptocamp.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+
+import logging
+from itertools import chain
+
+import werkzeug
+
+from odoo import http, models
+
+from ..registry import EndpointRegistry
+
+_logger = logging.getLogger(__name__)
+
+
+class IrHttp(models.AbstractModel):
+    _inherit = "ir.http"
+
+    @classmethod
+    def _endpoint_route_registry(cls, env):
+        return EndpointRegistry.registry_for(env.cr)
+
+    @classmethod
+    def _generate_routing_rules(cls, modules, converters):
+        # Override to inject custom endpoint rules.
+        return chain(
+            super()._generate_routing_rules(modules, converters),
+            cls._endpoint_routing_rules(),
+        )
+
+    @classmethod
+    def _endpoint_routing_rules(cls):
+        """Yield custom endpoint rules"""
+        e_registry = cls._endpoint_route_registry(http.request.env)
+        for endpoint_rule in e_registry.get_rules():
+            _logger.debug("LOADING %s", endpoint_rule)
+            endpoint = endpoint_rule.endpoint
+            for url in endpoint_rule.routing["routes"]:
+                yield (url, endpoint)
+
+    @classmethod
+    def routing_map(cls, key=None):
+        last_update = cls._get_routing_map_last_update(http.request.env)
```

Comment:
@simahawk To be efficient, we should not rely on the last_update but on the last value of a dedicated sequence. The computation of the last_update requires a query on the whole table followed by a sort on the last_update column. This kind of operation is far more resources consuming than a simple query to get the `last_value` from a sequence.